### PR TITLE
tools: Optimize drc math libraries

### PR DESCRIPTION
**Pull request by ShriramShastry**
Signed-off-by: sriram <malladi.sastry@intel.com>

For each of the  drc math library directory, there are folder  c_fixpt,matlab_fixpt,matlab_floatpt having code
-Usage and Q Format input and output information
Under each Math Lib directory 
- c fixpt code has initialization value under init_struc_fixpt().
- c fixpt routine init_struc_fixpt() receive input value/s are derived from matlab_fixpt_wrapper_fixt().
- matlab_fixpt_wrapper_fixt(input) is set from matlab_floatpt init_struc().

Testbench is designed for range of data , once library routines are verified, user can plug the c fixed point code to drc sof c code,

### CODE REVIEW

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
+ * File: drc_asin.h
```
Comment (johnylin76) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h:2]: Should we follow the rule to place header files to src/include/sof/audio?

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#define DRC_ASIN_H
+
+/* Include Files */
+#include <stddef.h>
+#include <stdlib.h>
+#include "typedef.h"
+#include "drc_asin_types.h"
+
+/* Function Declarations */
+extern void drc_asin_fixpt(const signed char x[2], int y[2]);
```
Comment (johnylin76) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h:18]: There should be a clear instruction comment to tell user what are the inputs and outputs, and their word length and fraction length, expected input and output value range, and etc.

For example, https://github.com/thesofproject/sof/blob/master/src/math/decibels.c#L74

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+  return z;
+}
+
+/*
+ * function y = drc_asin_fixpt(x)
+ * Arguments    : const signed char x[2]
+ *                int y[2]
+ * Return Type  : void
+ */
+void drc_asin_fixpt(const signed char x[2], int y[2])
```
Comment (johnylin76) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:136]: I would recommend to keep all c-fixpt function interfaces to always use int32_t input and output.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h [hunk 4]
```diff
@@ -0,0 +1,132 @@
+/*
+ * File: typedef.h
```
Comment (johnylin76) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h:2]: This file probably is able to be common among all c-fixpt implementations?

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h (hunk 2 above)
Comment (johnylin76) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h:18]: All other c-fixpt implementations as well.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+extern void drc_asin_initialize(void);
+extern void drc_asin_terminate(void);
+extern void init_struc_fixpt(signed char x[2]);
```
Comment (johnylin76) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h:21]: What do the above three (initialize, terminate, init_struc) need for?

File: src/audio/drc/drc_math_lib/asin/c_fixpt/main.c [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+{
+  (void)argc;
+  (void)argv;
+
+
+  //main_drc_asin_fixpt();
+  //main_init_struc_fixpt();
+
+  signed char x[2];
+  signed char iv[2];
```
Comment (johnylin76) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/main.c:29]: unused variable?

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+  signed char LUT[6]);
+
+/* Function Definitions */
+
+/*
+ * Arguments    : int u
+ *                unsigned int n
+ * Return Type  : int
+ */
+static int asr_s32(int u, unsigned int n)
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:23]: asr isn't clear to everyone, please provide a description. Please also apply to other functions.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+ *
+ * Author :Shriram Shastry
+ * Copyright (c) 2016, Intel Corporation All rights reserved.
+ */
+
+/* Include Files */
+#include "drc_asin.h"
+
+/* Function Declarations */
+static int asr_s32(int u, unsigned int n);
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:12]: please try to order your functions so you don't need forward declarations for static functions.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+  return y;
+}
+
+/*
+ * Arguments    : signed char sinValue
+ *                short nIters
+ *                const signed char LUT[6]
+ * Return Type  : signed char
+ */
+static signed char iScalarCordicAsin(signed char sinValue, short nIters, const
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:44]: i believe we use snake case.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+  signed char LUT[6])
+{
+  signed char z;
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:47]: I don't think I have seen signed char in this repo, int8_t might be a bit clearer.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+  else
+  {
+    y = (-((int)((unsigned int)(((unsigned int)((int)(-1 - u))) >> n)))) - 1;
+  }
+
+  return y;
+}
+
+/*
+ * Arguments    : signed char sinValue
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:39]: Please label units for these functions as I'm guessing the distinction between degrees and radians is relevant here. If possible please label variables used as well to help maintenance.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h (hunk 5 above)
Comment (cujomalainey) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h:21]: We should avoid extern if possible as well, what is the benefit of using that here?

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_types.h [hunk 12]
```diff
@@ -0,0 +1,19 @@
+/*
+ * File: drc_asin_types.h
+ *
+ * Author :Shriram Shastry
+ * Copyright (c) 2016, Intel Corporation All rights reserved.
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_types.h:5]: what is the point of this file?

File: src/audio/drc/drc_math_lib/asin/c_fixpt/main.h [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+#define MAIN_H
+
+/* Include Files */
+#include <stddef.h>
+#include <stdlib.h>
+#include "typedef.h"
+#include "drc_asin_types.h"
+
+/* Function Declarations */
+extern int main(int argc, const char * const argv[]);
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/main.h:19]: not sure why this file is needed? you don't need to forward declare main.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+
+
+/*=======================================================================*
+ * Fixed width word size data types:                                     *
+ *   int8_T, int16_T, int32_T     - signed 8, 16, or 32 bit integers     *
+ *   uint8_T, uint16_T, uint32_T  - unsigned 8, 16, or 32 bit integers   *
+ *   real32_T, real64_T           - 32 and 64 bit floating point numbers *
+ *=======================================================================*/
+typedef signed char int8_T;
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h:22]: why not just use inttypes.h?

File: src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+  uint32_T re;
+  uint32_T im;
+} cuint32_T;
+
+/*=======================================================================*
+ * Min and Max:                                                          *
+ *   int8_T, int16_T, int32_T     - signed 8, 16, or 32 bit integers     *
+ *   uint8_T, uint16_T, uint32_T  - unsigned 8, 16, or 32 bit integers   *
+ *=======================================================================*/
+#define MAX_int8_T                     ((int8_T)(127))
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h:103]: i think this is also defined in a system header as well

File: src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+#define MIN_uint16_T                   ((uint16_T)(0))
+#define MAX_int32_T                    ((int32_T)(2147483647))
+#define MIN_int32_T                    ((int32_T)(-2147483647-1))
+#define MAX_uint32_T                   ((uint32_T)(0xFFFFFFFFU))
+#define MIN_uint32_T                   ((uint32_T)(0))
+
+/* Logical type definitions */
+#if !defined(__cplusplus) && !defined(__true_false_are_keywords) && !defined(__bool_true_false_are_defined)
+#  ifndef false
+#   define false                       (0U)
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h:119]: I believe FALSE and TRUE are part of bool.h (or whatever the boolean header is)

File: src/audio/drc/drc_math_lib/asin/matlab_fixpt/drc_asin_wrapper_fixpt.m [hunk 17]
```diff
@@ -0,0 +1,13 @@
+function y = drc_asin_wrapper_fixpt(x)
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/matlab_fixpt/drc_asin_wrapper_fixpt.m:1]: what are units of x?

File: src/audio/drc/drc_math_lib/ceil/c_fixpt/drc_ceil_types.h [hunk 18]
```diff
@@ -0,0 +1,19 @@
+/*
+ * File: drc_ceil_types.h
+ *
+ * Author :Shriram Shastry
+ * Copyright (c) 2016, Intel Corporation All rights reserved.
+ */
+
+#ifndef DRC_CEIL_TYPES_H
+#define DRC_CEIL_TYPES_H
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/ceil/c_fixpt/drc_ceil_types.h:9]: empty header

File: src/audio/drc/drc_math_lib/ceil/c_fixpt/typedef.h [hunk 19]
```diff
@@ -0,0 +1,128 @@
+/*
+ * File: typedef.h
+ *
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/ceil/c_fixpt/typedef.h:3]: you have this file defined twice

File: src/audio/drc/drc_math_lib/asin/matlab_fixpt/init_struc_fixpt.m [hunk 20]
```diff
@@ -0,0 +1,13 @@
+function x = init_struc_fixpt ()
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/matlab_fixpt/init_struc_fixpt.m:1]: there should not be any matlab code under src/ unless there is a good reason, we should move this under toosl/tune

**Review by cujomalainey — CHANGES_REQUESTED** [review]
Left some initial comments, please dedeup files, fix style and comments and move code where needed across whole commit then I will re-review

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h (hunk 1 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h:2]: Yes, we should / must follow . Since it is a testbench code & deattached one, i didn't kept under/follow src/include/sof/audio

File: src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h (hunk 4 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h:2]: Yes

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h (hunk 2 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h:18]: OK - I'II add for all library function

File: src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h (hunk 4 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h:2]: Since each library function is a module in itself so typedef is added for every module. I agree typedef.c is common for all math library functions

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h (hunk 5 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.h:21]: I will remove initialize, terminate, function calls except  init_struc_fixpt(). function  init_struc_fixpt is initializing the testvector for library function.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/main.c (hunk 6 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/main.c:29]: signed char iv[2] will be removed

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c (hunk 7 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:23]: ASR : Arithmetic shift right  for 32 bit wordlength
    Arithmatic bit right shift is performed for the cordic iteration
    for angles less than pi/4, we initialize (x,y) along the
    x-axis[xShift,xDShift]. 
    For angles greater than or equal to pi/4, we initialize (x,y)
    along the y-axis[yShift,yDShift]. 
    This improves the accuracy of the algorithm near the edge of the domain of convergence.
    Variables j,k are derived from the Cordiac Iteration (WordLength-1)
     j = min(bitsll(i,1),maxShift)  // bitsll logical left shift
     k = min(i - 1,maxShift);

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c (hunk 9 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:44]: CORDIC-based approximation of inverse sine

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c (hunk 10 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:47]: OK - I will replace signed char with int8_t

File: src/audio/drc/drc_math_lib/asin/matlab_fixpt/init_struc_fixpt.m (hunk 20 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/matlab_fixpt/init_struc_fixpt.m:1]: This was done to have a testbench code  , I am ok to your suggestion, I will move the  all matlab floating and fixed scripts under tools/tune folder

File: src/audio/drc/drc_math_lib/asin/matlab_fixpt/drc_asin_wrapper_fixpt.m (hunk 17 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/matlab_fixpt/drc_asin_wrapper_fixpt.m:1]: x_in = fi( **x**, 1, 2, 0, fm );
**x** unit is represented by input values of -1 & 1
From **init_struc_fixpt.m** you  can see x units
In x = fi(**[-1,1]**, 1, 2, 0, fm); 
Representation expansion is 1 Signed bit,  2 Word Length and 0 Fraction Length along with
fm = fimath('RoundingMethod', 'Convergent',...
	     'OverflowAction', 'Wrap',...
	     'ProductMode','FullPrecision',...
	     'SumMode','FullPrecision');

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 21]
```diff
@@ -0,0 +1,221 @@
+/*
+ * File: drc_asin.c
```
Comment (lgirdwood) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:2]: Needs SPDX licence header - see the other files.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c (hunk 7 above)
Comment (lgirdwood) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:23]: @ShriramShastry that's good info above, but this needs to go into the comment above the function implementation as a doxygen comment. This way we can process all the API comments into an online developer guide.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+static int asr_s32(int u, unsigned int n)
+{
+  int y;
+  if (u >= 0)
+  {
+    y = (int)((unsigned int)(((unsigned int)u) >> n));
+  }
+  else
+  {
+    y = (-((int)((unsigned int)(((unsigned int)((int)(-1 - u))) >> n)))) - 1;
```
Comment (lgirdwood) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:32]: indentation - please use tabs (tab size is 8). Also please use std types like uint32_t or int16_t as we want this code to run exactly the same on 64bit x86  targets.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c (hunk 9 above)
Comment (lgirdwood) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:44]: @ShriramShastry not following, but the function naming is snake case i.e. scalar_cordic_asin()

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 23]
```diff
@@ -0,0 +1,221 @@
...
+  signed char y;
+  int i;
+  int b_i;
+  short j;
+  short k;
+  signed char xShift;
+  signed char xDShift;
+  signed char yShift;
+  signed char yDShift;
+  b_sinValue = (signed char)(sinValue * 32);
```
Comment (lgirdwood) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:59]: newline between every block.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 24]
```diff
@@ -0,0 +1,221 @@
...
+  if (b_sinValue > 22)
```
Comment (lgirdwood) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:60]: 22 looks like a magic number here. Best to use descriptive macros instead i.e.
```
/* this is used for XXX */
#define CORDIC_ASIN_SOME_USEFUL_VALUE 22

if (b_sin_value > CORDIC_ASIN_USEFUL_VALUE)
```

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 25]
```diff
@@ -0,0 +1,221 @@
...
+  y[1] = ((int)i) * 33554432;
+}
+
+/*
+ * Arguments    : void
+ * Return Type  : void
+ */
+void drc_asin_initialize(void)
+{
+}
```
Comment (lgirdwood) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:195]: Is this needed ?

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c [hunk 26]
```diff
@@ -0,0 +1,221 @@
...
+
+/*
+ * Arguments    : void
+ * Return Type  : void
+ */
+void drc_asin_terminate(void)
+{
+  /* (no terminate code required) */
+}
```
Comment (lgirdwood) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c:204]: ditto

File: src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h (hunk 14 above)
Comment (lgirdwood) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/typedef.h:22]: ack, @ShriramShastry please use stdint.h

**Review by lgirdwood — CHANGES_REQUESTED** [review]
Can we also have some comments within the functions adding some perspective for the reviewers about what each code block does in each function.

File: src/audio/drc/drc_math_lib/asin/matlab_fixpt/drc_asin_wrapper_fixpt.m (hunk 17 above)
Comment (cujomalainey) [reply src/audio/drc/drc_math_lib/asin/matlab_fixpt/drc_asin_wrapper_fixpt.m:1]: please add as a comment explaining

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c (hunk 10 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin.c]: OK

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c [hunk 27]
```diff
@@ -0,0 +1,234 @@
...
+
+
+/* Include Files */
+#include "drc_asin_fixpt.h"
+
+/* Type Definitions */
+#include "stdint.h"
+
+/* Function Declarations */
+static int32_T asr_s32(int32_T u, uint32_T n);
```
Comment (cujomalainey) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c:15]: these are still using the wrong types (lower case t)

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c [hunk 28]
```diff
@@ -0,0 +1,231 @@
...
+  return z;
+}
+
+ /*^^^
+ *-----------------------------------------------------------------------------
+ *
+ * Prototype        :void drc_asin_fixpt
+ * Description      :calculate fixed point inverse sine of x based on a CORDIC approximation
+ * Input Arguments  :const int8_T x[2] Signedness:1, Signed,WordLength: 2,FractionLength: 0
+ * Output Arguments :int32_T y[2] Q30.32   [QFL.WL]
```
Comment (johnylin76) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c:168]: The convention of presentation in SOF library should be Q(WL-FL).FL as I've seen elsewhere?
For example, Q30.32 here should be corrected to Q2.30
@lgirdwood please correct me if I'm wrong

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c [hunk 29]
```diff
@@ -0,0 +1,231 @@
...
+/* Function Definitions */
+
+ /*
+ *-----------------------------------------------------------------------------
+ *
+ * Prototype       :static scalar_cordic_asin
+ * Description     :Compute the inverse Sine using a CORDIC implementation.
+ * ASR 32          : Arithmetic shift right  for 32 bit
+ * Arithmetic bit right shift is performed for CORDIC iteration
+ * For angles less than pi/4, we initialize (x,y) along the x-axis [ see code variable - xShift,xDShift].
```
Comment (johnylin76) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c:28]: This feels like the overview comment, not for asr_s32?
If so, maybe put to the beginning of this file? And IMO this is a little bit too detailed.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/main.c [hunk 30]
```diff
@@ -9,6 +9,9 @@
 #include "main.h"
 #include "drc_asin_fixpt.h"
 
+/* Type Definitions */
+#include "stdint.h"
```
Comment (johnylin76) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/main.c:13]: This commit and the above one should be merged to the one above them?

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c (hunk 28 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c:168]: OK - I will change

File: src/audio/drc/drc_math_lib/asin/c_fixpt/main.c (hunk 30 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/main.c:13]: Yes, Correct, initially i missed adding in first time so made next commit.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c (hunk 27 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c:15]: OK - I change int32_T to lower case int32_t

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c (hunk 29 above)
Comment (ShriramShastry) [reply src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c:28]: This feels like the overview comment, not for asr_s32?
 I make change for  Description section,

File: src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c [hunk 31]
```diff
@@ -12,45 +12,40 @@
...
- * ASR 32          : Arithmetic shift right  for 32 bit
- * Arithmetic bit right shift is performed for CORDIC iteration
+ * Prototype       :static int32_t asr_s32
+ * Description     :perform Arithmetic bit right shift on cordic iteration.
  * For angles less than pi/4, we initialize (x,y) along the x-axis [ see code variable - xShift,xDShift].
  * For angles greater than or equal to pi/4, we initialize (x,y) along the y-axis[see code variable - yShift,yDShift].
  * This improves the accuracy of the algorithm near the edge of the domain of convergence.
  * Variables j,k are derived from the Cordiac Iteration ( 1 to wordLength-1)
  * j = min(bitsll(i,1),maxShift) // bitsll logical left shift
  * k = min(i - 1,maxShift);
```
Comment (johnylin76) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/drc_asin_fixpt.c:31]: Line 26~31 mainly describes the code implemented in scalar_cordic_asin(), right?
Maybe it is a better idea to have these comments along the code implementation, then it will be easier to follow the code and know what it does.

File: src/audio/drc/drc_math_lib/asin/c_fixpt/main.c [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+/* Include Files */
+#include "main.h"
+#include "drc_asin.h"
+
+/*
+ * Arguments    : int argc
+ *                const char * const argv[]
+ * Return Type  : int
+ */
+int main(int argc, const char * const argv[])
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/asin/c_fixpt/main.c]: If you need unit tests cmocka would be good I think. See SOF/test. Or to tools/testbench/? Also separate test stuff for separate commit.

File: src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c [hunk 33]
```diff
@@ -0,0 +1,463 @@
...
+
+  y[n1m1] = u1[n1m1] & ((1U << (32U - n2)) - 1U);
+}
+
+/*
+ * Arguments    : int32_T a
+ *                int32_T b
+ * Return Type  : int32_T
+ */
+static int32_T mul_s32_sat(int32_T a, int32_T b)
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c:53]: Use q_multsr_sat_32x32() from format.h?

File: src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c [hunk 34]
```diff
@@ -0,0 +1,463 @@
...
+}
+
+/*
+ * Arguments    : int32_T in0
+ *                int32_T in1
+ *                uint32_T *ptrOutBitsHi
+ *                uint32_T *ptrOutBitsLo
+ * Return Type  : void
+ */
+static void mul_wide_s32(int32_T in0, int32_T in1, uint32_T *ptrOutBitsHi,
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c:83]: Why not use int64_t and native multiply?

File: src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c [hunk 35]
```diff
@@ -0,0 +1,463 @@
...
+    {
+      absIn0++;
+    }
+  }
+
+  *ptrOutBitsHi = absIn0;
+  *ptrOutBitsLo = outBitsLo;
+}
+
+/*
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c:150]: It's hard to understand what these functions are doing without comments. Also the functions names are not descriptive, to me at least.

File: src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c [hunk 36]
```diff
@@ -0,0 +1,463 @@
...
+
+  while (i < n)
+  {
+    y[i] = 0U;
+    i++;
+  }
+}
+
+/*
+ * function y = drc_db2mag_fixpt(tstruct)
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c:370]: No description about algorithm, comments like SOF src/math/decibels.c math would be nice to understand how this algorithm works.

File: src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c [hunk 37]
```diff
@@ -0,0 +1,463 @@
...
+
+  /*  =power(10,(21/20)) */
+  return y;
+}
+
+/*
+ * Arguments    : void
+ * Return Type  : void
+ */
+void drc_db2mag_initialize(void)
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c:429]: Empty functions.

File: src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c [hunk 38]
```diff
@@ -0,0 +1,463 @@
...
+{
+  /* (no terminate code required) */
+}
+
+/*
+ * function [tstruct] = init_struc_fixpt
+ * Arguments    : struct0_T *tstruct
+ * Return Type  : void
+ */
+void init_struc_fixpt(struct0_T *tstruct)
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag.c:447]: Example of similar name function with different purpose here.

File: src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag_types.h [hunk 39]
```diff
@@ -0,0 +1,43 @@
...
+/* Type Definitions */
+#ifndef typedef_struct0_T
+#define typedef_struct0_T
+
+typedef struct
+{
+  uint8_T u1;
+  uint8_T u2;
+}
+struct0_T;
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag_types.h:23]: A very non-descriptive struct name. Also we don't use "typedef structs" in normal components, only some xtensa HAL codes contains such.

File: src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag_types.h [hunk 40]
```diff
@@ -0,0 +1,43 @@
...
+#endif                                 /*typedef_struct0_T*/
+
+#ifndef typedef_uint64m_T
+#define typedef_uint64m_T
+
+typedef struct
+{
+  uint32_T chunks[2];
+}
+uint64m_T;
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/db2mag/c_fixpt/drc_db2mag_types.h:34]: Use uint64_t ?

File: src/audio/drc/drc_math_lib/db2mag/c_fixpt/main.c [hunk 41]
```diff
@@ -0,0 +1,39 @@
...
+#include "main.h"
+#include "drc_db2mag.h"
+
+
+/*
+ * Arguments    : int32_T argc
+ *                const char * const argv[]
+ * Return Type  : int32_T
+ */
+int32_T main(int32_T argc, const char * const argv[])
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/db2mag/c_fixpt/main.c:19]: To cmocka or testbench?

File: src/audio/drc/drc_math_lib/db2mag/c_fixpt/typedef.h [hunk 42]
```diff
@@ -0,0 +1,132 @@
+/*
+ * File: typedef.h
+ *
+ * Author :Shriram Shastry
+ * Copyright (c) 2016, Intel Corporation All rights reserved.
+ */
+
+#ifndef TYPEDEF_H
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/db2mag/c_fixpt/typedef.h:8]: Deja-vu?

File: src/audio/drc/drc_math_lib/db2mag/matlab_fixpt/drc_db2mag_fixpt.m [hunk 43]
```diff
@@ -0,0 +1,27 @@
+function y = drc_db2mag_fixpt(tstruct)
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/db2mag/matlab_fixpt/drc_db2mag_fixpt.m]: Is this Matlab coder source file? A new directory e.g. tools/reference/<your work>/ might be suitable.

File: src/audio/drc/drc_math_lib/floor/c_fixpt/drc_floor.c [hunk 44]
```diff
@@ -0,0 +1,53 @@
...
+ *                uint8_T y[5]
+ * Return Type  : void
+ */
+void drc_floor_fixpt(const uint32_T x[5], uint8_T y[5])
+{
+  int32_T i;
+
+  for (i = 0; i < 5; i++)
+  {
+    y[i] = (uint8_T)(x[i] >> ((uint32_T)31));
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/floor/c_fixpt/drc_floor.c:25]: Bytes are not recommended and why is 31 typecasted?

File: src/audio/drc/drc_math_lib/floor/c_fixpt/drc_floor.c [hunk 45]
```diff
@@ -0,0 +1,53 @@
...
+ * function x = init_struc_fixpt
+ * Arguments    : uint32_T x[5]
+ * Return Type  : void
+ */
+void init_struc_fixpt(uint32_T x[5])
+{
+  int32_T i;
+  static const uint32_T uv[5] =
+  {
+    3321224261U, 3535972626U, 3750720991U, 3965469356U, 4180217721U
```
Comment (singalsu) [inline src/audio/drc/drc_math_lib/floor/c_fixpt/drc_floor.c:39]: Magic numbers, what are these?

**Review by singalsu — CHANGES_REQUESTED** [review]

**Comment by singalsu** [comment]
You should apply the PR updates to appropriate commits to avoid us reviewers to consume a lot of time for things you have changed in another commit. Use the interactive rebase "git rebase -i @~N" where N is number of your commits to edit edit the proper patch.

**Comment by paulstelian97** [comment]
@ShriramShastry Did the close was on purpose? Github will close pull requests if you force push and none of the original comments could be found.

**Comment by ShriramShastry** [comment]
> @ShriramShastry Did the close was on purpose? Github will close pull requests if you force push and none of the original comments could be found.

No - I did not close it.  May be accidental . I am still in the process of adding code . Waiting for @cujomalainey  to complete  testbench framework completion.

**Comment by cujomalainey** [comment]
We can leave this closed until we get the code into working condition in @ShriramShastry repo
